drivers/at_cmd/at_cmd.c

```
@@ -0,0 +1,299 @@
+			break;
+		}
+	} while (0);
+
+	if ((return_code == AT_CMD_ERROR_CMS) ||
+	    (return_code == AT_CMD_ERROR_CME)) {
+		last_error = atoi(&buf[ARRAY_SIZE(AT_CMD_CMS_STR) - 1]);
+	} else {
+		last_error = 0;
+	}
+
+	if (tmpstr) {
+		buf[tmpstr - buf] = '\0';
+	}
+
+	return return_code;
+}
+
+static void callback_worker(struct k_work *item)
+{
+	struct callback_work_item *data =
+		CONTAINER_OF(item, struct callback_work_item, work);
+
+	if ((data != NULL) && (data->callback != NULL)) {
+		data->callback(data->data);
+	}
+
+	k_free(data);
+}
+
+
+static void socket_thread_fn(void *arg1, void *arg2, void *arg3)
+{
+	int bytes_read;
+	char buf[RESPONSE_MAX_LEN];
+	enum at_cmd_rc return_code;
+	bool write_with_response = false;
+
+	ARG_UNUSED(arg1);
+	ARG_UNUSED(arg2);
+	ARG_UNUSED(arg3);
+
+	LOG_DBG("AT socket thread started");
+
+	for (;;) {
+		bytes_read = recv(common_socket_fd, buf, sizeof(buf), 0);
+		write_with_response = false;
+
+		if (bytes_read < 0) {
+			LOG_ERR("AT socket recv failed with err %d",
+				bytes_read);
+
+			if ((close(common_socket_fd) == 0) &&
+			    (open_socket() == 0)) {
+				LOG_INF("AT socket recovered");
+
+				if (cmd_pending.lock_count > 0) {
+					return_code = AT_CMD_ERROR;
+					goto next;
+				} else {
+					continue;
+				}
+			}
+
+			LOG_ERR("Unrecoverable reception error (err: %d), "
+				"thread killed", errno);
+			return;
+		} else if (bytes_read == sizeof(buf) ||
+			   buf[bytes_read - 1] != '\0') {
+
+			LOG_ERR("AT message to large for reception buffer or "
+				"missing termination character");
+			return_code = AT_CMD_ERROR;
+			goto next;
+		}
+
+		return_code = get_return_code(buf);
+
+		if (return_code != AT_CMD_NOTIFICATION) {
+			if ((response_buf_len > 0) &&
+			    (response_buf != NULL)) {
+				if (response_buf_len > strlen(buf)) {
+					strncpy(response_buf,
+						buf,
+						response_buf_len);
+				} else {
+					LOG_ERR("Response buffer not large "
+						"enough");
+					response_buf = NULL;
```

You do not need this line.

---

As discussed, you should return an explicit error if this happens.